### PR TITLE
Add more Llama API

### PR DESCRIPTION
**Pull request by BillyGalbreath**
This adds some useful API to Llamas, including two new events

### CODE REVIEW

File: patches/server/0827-Add-more-Llama-API.patch [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+         this.caravanHead = llama;
+         this.caravanHead.caravanTail = this;
+     }
+@@ -450,6 +457,13 @@ public class Llama extends AbstractChestedHorse implements RangedAttackMob {
+         return this.caravanHead;
+     }
+ 
++    // Paper start
++    @Nullable
++    public Llama getCaravanTail() {
```
Comment (Machine-Maker) [inline patches/server/0827-Add-more-Llama-API.patch]: Probably better to just access transform the field instead of adding a getter for it

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

File: patches/api/0338-Add-more-Llama-API.patch [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
++}
+diff --git a/src/main/java/org/bukkit/entity/Llama.java b/src/main/java/org/bukkit/entity/Llama.java
+index d23226ccb0f6c25028f000ce31346cd0a8898e6a..d0057be88f7838449839e51d0a0e93dea6284bb7 100644
+--- a/src/main/java/org/bukkit/entity/Llama.java
++++ b/src/main/java/org/bukkit/entity/Llama.java
+@@ -3,6 +3,7 @@ package org.bukkit.entity;
+ import com.destroystokyo.paper.entity.RangedEntity;
+ import org.bukkit.inventory.LlamaInventory;
+ import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
```
Comment (lynxplay) [inline patches/api/0338-Add-more-Llama-API.patch]: Should probably be replaced either with its fully qualified name in the respective usecases (tho the ranged entity api imports a few lines above seems to disagree with the guidelines :joy:)

File: patches/api/0338-Add-more-Llama-API.patch [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
++     * Leave current caravan if in one
++     */
++    void leaveCaravan();
++
++    /**
++     * Check if another Llama is following this Llama
++     *
++     * @return True if being followed in the caravan
++     */
++    boolean hasCaravanTail();
```
Comment (lynxplay) [inline patches/api/0338-Add-more-Llama-API.patch:162]: Is this method actually useful ? I know it maps an internal method directly to the API but it seems to be just a somewhat useful shortcut for `getCaravanTail() != null`.

File: patches/server/0827-Add-more-Llama-API.patch [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+     public Llama caravanTail;
++    public boolean shouldJoinCaravan = true; // Paper
+ 
+     public Llama(EntityType<? extends Llama> type, Level world) {
+         super(type, world);
+@@ -95,7 +96,7 @@ public class Llama extends AbstractChestedHorse implements RangedAttackMob {
+         if (!this.inventory.getItem(1).isEmpty()) {
+             nbt.put("DecorItem", this.inventory.getItem(1).save(new CompoundTag()));
+         }
+-
```
Comment (Proximyst) [inline patches/server/0827-Add-more-Llama-API.patch]: Don't drop this line, please. Same applies for other places where such is done.

**Review by Proximyst — APPROVED** [review]
Looks OK to me.

File: patches/api/0338-Add-more-Llama-API.patch (hunk 3 above)
Comment (Proximyst) [reply patches/api/0338-Add-more-Llama-API.patch:162]: It's common practice to have this method in the API for some reason. I don't find it useful personally, but it's not a big deal to just have it.

Note also that it's not even useful in terms of IDE (and Kotlin) support for it, as an IDE will still yell at:

```java
if (!llama.hasCaravanTail()) {
  return;
}

llama.getCaravanTail().isOp(); // IntelliJ will say it's potentially throwing an NPE.
```

**Review by Machine-Maker — APPROVED** [review]

**Review by MiniDigger — APPROVED** [review]

**Pull request closed without merging**